membership_product_set/tests/test_membership.py

```diff
@@ -0,0 +1,127 @@
+        cls.tax_15p = cls.env['account.tax'].create({
+            'type_tax_use': 'sale',
+            'amount_type': 'percent',
+            'name': "Tax 15%",
+            'amount': 15,
+            'active': True,
+        })
+        cls.account_partner_type = cls.env['account.account.type'].create({
+            'name': 'Test partner account type',
+            'type': 'receivable',
+        })
+        cls.account_partner = cls.env['account.account'].create({
+            'name': 'Test partner account',
+            'code': 'PARTNER',
+            'user_type_id': cls.account_partner_type.id,
+            'reconcile': True,
+        })
+        cls.partner = cls.env['res.partner'].create({
+            'name': 'Test partner',
+            'free_member': True,
+            'property_account_receivable_id': cls.account_partner.id,
+        })
+        cls.partner_wo_account = cls.env['res.partner'].create({
+            'name': 'Test Contact WO Account',
+            'property_account_receivable_id': False,
+        })
+        cls.gold_product = cls.env['product.product'].create({
+            'type': 'service',
+            'name': 'Membership Gold',
+            'membership': True,
+            'membership_date_from': fields.Date.today(),
+            'membership_date_to': cls.next_month,
+            'list_price': 100.00,
+            'taxes_id': [(4, cls.tax_15p.id, False)]
+        })
+        cls.silver_product = cls.env['product.product'].create({
+            'type': 'service',
+            'name': 'Membership Silver',
+            'membership': True,
+            'membership_date_from': fields.Date.today(),
+            'membership_date_to': cls.next_month,
+            'list_price': 50.00,
+        })
+        cls.basic_product = cls.env['product.product'].create({
+            'type': 'service',
+            'name': 'Membership Basic',
+            'membership': True,
+            'membership_date_from': fields.Date.today(),
+            'membership_date_to': cls.next_month,
+            'list_price': 20.00,
+        })
+        cls.product_set = cls.env['product.product'].create({
+            'type': 'service',
+            'name': 'Membership Set',
+            'membership': True,
+            'membership_set': True,
+            'membership_date_from': fields.Date.today(),
+            'membership_date_to': cls.next_month,
+            'list_price': 150.00,
+            'membership_set_products': [
+                (4, cls.gold_product.id, False),
+                (4, cls.silver_product.id, False),
+            ]
+        })
+
+    def test_membership_invoice(self):
+        wizard = self.env['membership.invoice'].with_context(
+            active_ids=[self.partner.id],
+            active_model='res.partner'
+        ).create({
+            'product_id': self.product_set.id,
+            'member_price': 150,
+        })
+        with self.assertRaises(UserError):
```

Comment:
Consider using  **assertRaisesRegexp**, it can avoid to catch another **UserError** not intended and swallow the failure. 

` with self.assertRaisesRegexp(UserError, r'^My expected message')`

Reply:
Thanks for the tip. I will improve it now.